### PR TITLE
Added support for arrays of types with TryParse, StringValues and string[] for query strings and headers

### LINKED ISSUE #32516: Minimal Actions - array support in method signatures from querystring binding
### Describe the bug
When using the ASPP.NET 6 minimal hosting, querystring binding does not work with arrays/lists

### To Reproduce
```csharp

app.MapGet("/qs/{id:int}", HttpHandlers.GetByQueryString);

public static string GetByQueryString(int id, string name, int[] values)
{
    return id + " " + name + string.Join(",", values);
}
```

Adding `[FromQuery(Name = "values")]` to the method signature does not help either.

### Exceptions (if any)
An exception is thrown when sending in a request:

`curl http://localhost:5000/qs/123\?name\=dave\&values\=1\&values\=2`

`Unhandled exception. System.InvalidOperationException: No public static bool Int32[].TryParse(string, out Int32[]) method found for values.`

### Further technical details
- ASP.NET Core version 6
```
.NET SDK (reflecting any global.json):
 Version:   6.0.100-preview.5.21252.2
 Commit:    921395c33e

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  11.3
 OS Platform: Darwin
 RID:         osx.11.0-x64
 Base Path:   /Users/jonathan/.dotnet/sdk/6.0.100-preview.5.21252.2/

Host (useful for support):
  Version: 6.0.0-preview.5.21251.4
  Commit:  d2fba8fdc4

.NET SDKs installed:
  2.1.809 [/Users/jonathan/.dotnet/sdk]
  3.1.401 [/Users/jonathan/.dotnet/sdk]
  3.1.403 [/Users/jonathan/.dotnet/sdk]
  5.0.103 [/Users/jonathan/.dotnet/sdk]
  6.0.100-preview.5.21229.26 [/Users/jonathan/.dotnet/sdk]
  6.0.100-preview.5.21252.2 [/Users/jonathan/.dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.21 [/Users/jonathan/.dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.21 [/Users/jonathan/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.7 [/Users/jonathan/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.9 [/Users/jonathan/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.3 [/Users/jonathan/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.5.21229.9 [/Users/jonathan/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.5.21230.5 [/Users/jonathan/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.21 [/Users/jonathan/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.7 [/Users/jonathan/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.9 [/Users/jonathan/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.3 [/Users/jonathan/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.5.21229.1 [/Users/jonathan/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.5.21251.4 [/Users/jonathan/.dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download

```

### LINKED ISSUE #36726: Minimal APIs doesn't support binding to StringValues from query or headers
### Describe the bug
Parameter binding doesn't support StringValues (which is ASP.NET Core's native representation of both query string and header values).

### To Reproduce

```C#
using Microsoft.Extensions.Primitives;

var builder = WebApplication.CreateBuilder(args);

var app = builder.Build();

app.MapGet("/", (StringValues query) => query.ToString());

app.Run();
```

**OR**

```C#
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Primitives;

var builder = WebApplication.CreateBuilder(args);

var app = builder.Build();

app.MapGet("/", ([FromQuery] StringValues query) => query.ToString());

app.Run();
```

### Exceptions (if any)

The first fails parameter binding things that `StringValues` is a body parameter.

```
Unhandled exception. System.InvalidOperationException: Body was inferred but the method does not allow inferred body parameters.
Below is the list of parameters that we found:

Parameter           | Source
---------------------------------------------------------------------------------
query               | Body (Inferred)

Did you mean to register the "Body (Inferred)" parameter(s) as a Service or apply the [FromService] or [FromBody] attribute?
```

When adding `[FromQuery]`, it also fails but it's because `StringValues` doesn't have a `TryParse` method:

```
Unhandled exception. System.InvalidOperationException: No public static bool StringValues.TryParse(string, out StringValues) method found for query.
```

It feels bad that our native representation doesn't work...

### PR DESCRIPTION
**Pull request by davidfowl**
- Added inferred query support for methods where the body inference is disabled.
- Added a tests.

Fixes #32516 
Fixes #36726

### CODE REVIEW

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -342,6 +349,17 @@ private static Expression CreateArgument(ParameterInfo parameter, FactoryContext
             factoryContext.TrackedParameters.Add(parameter.Name, RequestDelegateFactoryConstants.RouteOrQueryStringParameter);
             return BindParameterFromRouteValueOrQueryString(parameter, parameter.Name, factoryContext);
         }
+        else if (parameter.ParameterType.IsArray && factoryContext.DisableInferredFromBody && ParameterBindingMethodCache.HasTryParseMethod(parameter.ParameterType.GetElementType()!))
+        {
+            // We only infer parameter types if you have an array of TryParsables and:
+            // - DisableInferredFromBody is true
+            // - TBD: You have explicitly declared a [FromBody] so it's no longer ambiguous
```
Comment (davidfowl) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:356]: Not sure about this but its an idea.

**Comment by davidfowl** [comment]
> Say for example that a query string contained values like 1,2,,4 because there's a blank/missing entry. Would it be explicitly ignored because empty (so you'd get new int[] { 1, 2, 4 }), would it fail the binding because the int.TryParse() call would return false, or could the user opt into one behaviour vs. the other?

This is a good question, currently I think it would fail to bind but I could go either way here (ignoring empty entries).

> Would this support query strings in the formats a=1,2,3, a=1&a=2&a=3, or both?

No, it will only support the second, not the first. The one thing I don't want to support multiple formats unless they are common. MVC supports LOTS of formats https://docs.microsoft.com/en-us/aspnet/core/mvc/models/model-binding?view=aspnetcore-6.0#collections, that won't be supported here. I chose the most obvious one (the latter in your above sample).

**Comment by martincostello** [comment]
> No, it will only support the second, not the first.

That's fair - I figured I'd ask the question about how much of MVC's leniency would come through as I've seen both approaches in the wild.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 2]
```diff
@@ -970,36 +1012,79 @@ private static Expression BindParameterFromValue(ParameterInfo parameter, Expres
...
+        // We can reuse this like we reuse temp source string
+        ParameterExpression? stringArrayExpr = parameter.ParameterType.IsArray ? Expression.Variable(typeof(string[]), "tempStringArray") : null;
 
-        var fullParamCheckBlock = !isOptional
-            ? Expression.Block(
+        var fullParamCheckBlock = (parameter.ParameterType.IsArray, isOptional) switch
+        {
+            (true, _) =>
+
+            Expression.Block(
```
Comment (davidfowl) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:1037]: This needs to handle nullability. It currently doesn't.

**Comment by davidfowl** [comment]
> That's fair - I figured I'd ask the question about how much of MVC's leniency would come through as I've seen both approaches in the wild.

If this is common, we can support it. I'll let others chime in of course 😄. I'm inclined to support a subset of whatever MVC supports. I just tried it an MVC does support it so its on the table.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 3]
```diff
@@ -202,6 +202,13 @@ private static Expression[] CreateArguments(ParameterInfo[]? parameters, Factory
             var errorMessage = BuildErrorMessageForInferredBodyParameter(factoryContext);
             throw new InvalidOperationException(errorMessage);
         }
+
+        if (factoryContext.HasInferredBody && factoryContext.HasInferredQueryArray)
+        {
+            var errorMessage = BuildErrorMessageForQueryAndJsonBodyParameters(factoryContext);
+            throw new InvalidOperationException(errorMessage);
```
Comment (halter73) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:209]: Is this currently possible considering `factoryContext.DisableInferredFromBody` must be true for `factoryContext.HasInferredBody` to be set? What if `[FromQuery]` is explicitly used?

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs (hunk 1 above)
Comment (halter73) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:356]: It's an interesting idea. I don't love one parameter influencing the binding source of another. I like keeping the binding source inference rules as simple as possible.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 4]
```diff
@@ -940,8 +960,30 @@ private static Expression BindParameterFromValue(ParameterInfo parameter, Expres
...
+        //         if (int.TryParse(tempSourceString, out var parsedValue))
+        //         {
+        //             param_local[i] = parsedValue;
+        //         }
+        //         else
+        //         {
+        //             wasParamCheckFailure = true;
+        //             Log.ParameterBindingFailed(httpContext, "Int32[]", "param1", tempSourceString);
+        //             break;
+        //         }
```
Comment (halter73) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:981]:
```suggestion
        //         }
        //         index++;
```

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 5]
```diff
@@ -940,8 +960,30 @@ private static Expression BindParameterFromValue(ParameterInfo parameter, Expres
         //     param2_local = 42;
         // }
 
+        // string[]? values = httpContext.Request.Query["param1"].ToArray();
```
Comment (halter73) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:963]: Is ToArray() actually being called or are StringValues being used directly?

Comment (halter73) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:963]: I see. It's an implicit conversion.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 6]
```diff
@@ -1459,13 +1544,14 @@ private class FactoryContext
         public bool AllowEmptyRequestBody { get; set; }
 
         public bool UsingTempSourceString { get; set; }
-        public List<ParameterExpression> ExtraLocals { get; } = new();
-        public List<Expression> ParamCheckExpressions { get; } = new();
+        public List<ParameterExpression> Locals { get; } = new();
+        public List<Expression> Expressions { get; } = new();
```
Comment (halter73) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:1548]: Why did you rename this? All the expressions in this list are still doing some sort of parameter validation, right?

Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:1548]: Because the names are bad. These aren't really about checking parameters its just a big old block being built with locals and statements.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs (hunk 3 above)
Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:209]: I think you're right, I would need to change the order to get a "better exception" but maybe it isn't worth it?

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs (hunk 1 above)
Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:356]: Ya, the more I thought about the more I think we should keep the rules simple. Implementing it would require changing how we bound the arguments doing code generation in a second pass after looking at all of the arguments.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs (hunk 6 above)
Comment (halter73) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:1548]: I'd buy that if and when we use this for general expressions, but the new naming implies that these are all the expressions being compiled even though most of the expressions (those returned by `CreateArgument()`) don't end up in this list.

`ParamCheckExpressions` is only used by `CreateParamCheckingResponseWritingMethodCall` to include additional expressions that all include `Expression.Assign(WasParamCheckFailureExpr, Expression.Constant(true))`. We check `factoryContext.ParamCheckExpressions.Count > 0` to determine if we need to add the following to the returned `RequestDelegate`:

```C#
Expression.Condition(WasParamCheckFailureExpr,
            set400StatusAndReturnCompletedTask,
            AddResponseWritingToMethodCall(methodCall, methodInfo.ReturnType))
```

This makes absolutely no sense unless all the expressions that could end up in the list are doing parameter checking. I agree this mechanism could be expanded and used more generally with further modification (mainly using some other state to determine if we're doing param checking), but I don't see the point of removing the self-documentation until then.

Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs]: Got it, I reverted the name change.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 7]
```diff
@@ -342,6 +349,17 @@ private static Expression CreateArgument(ParameterInfo parameter, FactoryContext
             factoryContext.TrackedParameters.Add(parameter.Name, RequestDelegateFactoryConstants.RouteOrQueryStringParameter);
             return BindParameterFromRouteValueOrQueryString(parameter, parameter.Name, factoryContext);
         }
+        else if (parameter.ParameterType.IsArray && factoryContext.DisableInferredFromBody && ParameterBindingMethodCache.HasTryParseMethod(parameter.ParameterType.GetElementType()!))
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:352]: Is `string[]` handled elsewhere? Since `string` doesn't have a `TryParse` it would skip this `if`

Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:352]: Haven't done it yet.

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs [hunk 8]
```diff
@@ -68,6 +69,17 @@ public void OnProvidersExecuted(ApiDescriptionProviderContext context)
 
     private ApiDescription CreateApiDescription(RouteEndpoint routeEndpoint, string httpMethod, MethodInfo methodInfo)
     {
+        static bool ShouldDisableInferredBody(string method)
```
Comment (davidfowl) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:72]: More copied logic between Map and the api description provider 😢

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 9]
```diff
@@ -451,6 +451,56 @@ void TestAction([FromRoute] int foo)
...
+                    new object[] { (Action<HttpContext, AddressFamily[]>)Store, new[] { "Unix" },new[] { AddressFamily.Unix } },
+                    new object[] { (Action<HttpContext, ILOpCode[]>)Store, new[] { "Nop" }, new[] { ILOpCode.Nop } },
+                    new object[] { (Action<HttpContext, AssemblyFlags[]>)Store, new[] { "PublicKey,Retargetable" },new[] { AssemblyFlags.PublicKey | AssemblyFlags.Retargetable } },
+                    new object[] { (Action<HttpContext, int?[]>)Store, new[] { "42" }, new int?[] { 42 } },
+                    new object[] { (Action<HttpContext, MyEnum[]>)Store, new[] { "ValueB" },new[] { MyEnum.ValueB } },
+                    new object[] { (Action<HttpContext, MyTryParseRecord[]>)Store, new[] { "https://example.org" },new[] { new MyTryParseRecord(new Uri("https://example.org")) } },
+                    new object?[] { (Action<HttpContext, int?[]>)Store, null, null },
+                    new object?[] { (Action<HttpContext, int[]>)Store, new string[] {}, Array.Empty<int>() },
+                    new object?[] { (Action<HttpContext, int?[]>)Store, new string[] { "1", "2", "", "4" }, new int?[] { 1,2, null, 4 } },
```
Comment (davidfowl) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:499]: This is the final case that needs to be handled.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs (hunk 7 above)
Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs]: Done now!

**Comment by davidfowl** [comment]
OK this is done and ready for review.

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 10]
```diff
@@ -451,6 +451,58 @@ void TestAction([FromRoute] int foo)
...
+                    new object[] { (Action<HttpContext, IPEndPoint[]>)Store, new[] { "127.0.0.1:80" },new[] { new IPEndPoint(IPAddress.Loopback, 80) } },
+                    new object[] { (Action<HttpContext, AddressFamily[]>)Store, new[] { "Unix" },new[] { AddressFamily.Unix } },
+                    new object[] { (Action<HttpContext, ILOpCode[]>)Store, new[] { "Nop" }, new[] { ILOpCode.Nop } },
+                    new object[] { (Action<HttpContext, AssemblyFlags[]>)Store, new[] { "PublicKey,Retargetable" },new[] { AssemblyFlags.PublicKey | AssemblyFlags.Retargetable } },
+                    new object[] { (Action<HttpContext, int?[]>)Store, new[] { "42" }, new int?[] { 42 } },
+                    new object[] { (Action<HttpContext, MyEnum[]>)Store, new[] { "ValueB" },new[] { MyEnum.ValueB } },
+                    new object[] { (Action<HttpContext, MyTryParseRecord[]>)Store, new[] { "https://example.org" },new[] { new MyTryParseRecord(new Uri("https://example.org")) } },
+                    new object?[] { (Action<HttpContext, int?[]>)Store, null, null },
```
Comment (halter73) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:497]: Is this the right expected behavior? Isn't the entire parameter being `null` only valid if the parameter is `int[]?` or `int?[]?`?

Comment (davidfowl) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:497]: Wait, I need to write another test, its likely that it does 400, but this test doesn't distinguish between the data being set to null and the handler never being called.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 11]
```diff
@@ -970,33 +1010,104 @@ private static Expression BindParameterFromValue(ParameterInfo parameter, Expres
...
+                                 ),
+                                 // index++
+                                 Expression.PostIncrementAssign(index)
+                            )
+                        , loopExit)
+                    ) : null;
+
+        var fullParamCheckBlock = (parameter.ParameterType.IsArray, isOptional) switch
+        {
+            // (isArray: false, optional: true)
```
Comment (davidfowl) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:1064]:
```suggestion
            // (isArray: true, optional: true)
```

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 12]
```diff
@@ -970,33 +1010,104 @@ private static Expression BindParameterFromValue(ParameterInfo parameter, Expres
...
+                new[] { index, stringArrayExpr! },
+                // values = httpContext.Request.Query["id"];
+                Expression.Assign(stringArrayExpr!, valueExpression),
+                Expression.IfThen(
+                    Expression.NotEqual(stringArrayExpr!, Expression.Constant(null)),
+                    arrayLoop!
+                )
+            ),
+
+            // (isArray: false, optional: false)
```
Comment (davidfowl) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:1077]:
```suggestion
            // (isArray: true, optional: false)
```

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 13]
```diff
@@ -980,6 +1126,76 @@ void TestAction([FromRoute] int tryParsable, [FromRoute] int tryParsable2)
...
+    }
+
+    [Fact]
+    public async Task RequestDelegateThrowsForTryParsableFailuresIfThrowOnBadRequestWithNonOptionalArrays()
+    {
+        var invoked = false;
+
+        static void StoreNullableIntArray(HttpContext httpContext, int?[] values)
+        {
+            httpContext.Items["values"] = values;
```
Comment (halter73) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:1171]: Nit. This test doesn't check `httpContext.Items`, but it does check `invoked`.

```suggestion
            invoked = true;
```

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs (hunk 8 above)
Comment (halter73) [reply src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:72]: Can we move this to shared source to prevent these from going out of sync? Maybe rename `ParameterBindingMethodCache` to be more general. If not, I'd say just add a comment to the `ShouldDisableInferredBody(string method)` in `EndpointRouteBuilderExtensions` to keep these in sync.

File: src/Mvc/Mvc.ApiExplorer/test/EndpointMetadataApiDescriptionProviderTest.cs [hunk 14]
```diff
@@ -329,16 +330,21 @@ static void AssertPathParameter(ApiDescription apiDescription)
...
 
-        AssertQueryParameter(GetApiDescription((int foo) => { }, "/"));
-        AssertQueryParameter(GetApiDescription(([FromQuery] int foo) => { }));
+        AssertQueryParameter<int>(GetApiDescription((int foo) => { }, "/"));
+        AssertQueryParameter<int>(GetApiDescription(([FromQuery] int foo) => { }));
+        AssertQueryParameter<TryParseStringRecordStruct>(GetApiDescription(([FromQuery] TryParseStringRecordStruct foo) => { }));
+        AssertQueryParameter<int[]>(GetApiDescription((int[] foo) => { }, "/"));
+        AssertQueryParameter<string[]>(GetApiDescription((string[] foo) => { }, "/"));
+        AssertQueryParameter<StringValues>(GetApiDescription((StringValues foo) => { }, "/"));
+        AssertQueryParameter<TryParseStringRecordStruct[]>(GetApiDescription((TryParseStringRecordStruct[] foo) => { }, "/"));
```
Comment (halter73) [inline src/Mvc/Mvc.ApiExplorer/test/EndpointMetadataApiDescriptionProviderTest.cs:347]: GetApiDescription defaults to "GET", can we add a test case for the non-ShouldDisableInferredBody()-case?

**Review by halter73 — APPROVED** [review]

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs (hunk 8 above)
Comment (halter73) [reply src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:72]: Oh, I just noticed that this should be called against all the methods in `IHttpMethodMetadata.HttpMethods`, not just the one we're creating the ApiDescription for currently to align with the logic in `EndpointRouteBuilderExtensions`.
